### PR TITLE
FW hand-launch takeoff improvements

### PR DESCRIPTION
**Pull request by sfuhrer**
Further FW improvements based on https://github.com/PX4/PX4-Autopilot/pull/20537, mostly centered around automatic hand-launch. 

## Describe problem solved by this pull request
- bad FW land detection and complicated/impossible to customize
- launch detection is messed up if before launch detection the land detection triggers (to !landed)

## Describe your solution
- link land detection state machine to launch detection: don't allow the land detection to go to !landed as long as no launch is detected (only when launch detection is enabled)
- expose land detection triggering time as parameter (LNDFW_TRIG_TIME)
- reduce default of LNDFW_VEL_Z_MAX to 1m/s (it's unlikely that a vehicle sitting on the ground has a larger v_z estimate). And remove the scale of 0.7 that was applied in the airspeed-less case to make the thresholds only tighter then.
- remove LAUN_CAT_PMAX parameter that was only during the motor delay phase of a catapult start. According to the param description, it is only there to limit the max pitch up during a bungee launch, and beside that I'm doubting that many people use PX4 on bungee launches, I don't see it crucial for safe operation. 
- set the bearing setpoint during a handlaunch to the yaw at launch detection for NPFG

While on it I couldn't hold back from refactoring the whole launch detection architecture quite a bit. The original intend (asaik) was to have different launch methods implemented, that then each have their own parameters and state machine. Implemented was though only one (catapult, which also works for hand-launch). So I removed the rather complex setup with dynamically instantiated methods and limit the launch detection to catapult, which results in having a very compact and easy to maintain state machine and structure. 

## Test data / coverage
SITL tested and one real flight.

### CODE REVIEW

**Review by tstastny — CHANGES_REQUESTED** [review]
Thanks a lot for the refactor!! 

One more small thing, aside from the review comments below (couldn't add comment to lines that weren't changed in the commits):

This is not related to your PR .. but while we're making things nicer, I noticed this variable name wasn't the most clear, could we make it e.g. `_vehicle_local_position`?
https://github.com/PX4/PX4-Autopilot/blob/3b5b585841ba482add3e89f4937552efa242f9d1/src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp#L1407

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 1]
```diff
@@ -1586,26 +1585,22 @@ FixedwingPositionControl::control_auto_takeoff(const hrt_abstime &now, const flo
 
 			/* Select throttle: only in LAUNCHDETECTION_RES_DETECTED_ENABLEMOTORS we want to use
 			 * full throttle, otherwise we use idle throttle */
```
[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:1587]: removed.

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 2]
```diff
@@ -1586,26 +1585,22 @@ FixedwingPositionControl::control_auto_takeoff(const hrt_abstime &now, const flo
...
-			const float max_takeoff_throttle = (_launch_detection_state != LAUNCHDETECTION_RES_DETECTED_ENABLEMOTORS) ?
+			const float max_takeoff_throttle = (_launchDetector.getLaunchDetected() < launch_detection_status_s::STATE_FLYING) ?
```
[the remark this answers is not in the document]
Comment (tstastny) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:1588]: nvm

File: src/modules/land_detector/FixedwingLandDetector.cpp [hunk 3]
```diff
@@ -60,7 +60,16 @@ bool FixedwingLandDetector::_get_landed_state()
 
 	bool landDetected = false;
 
-	if (hrt_elapsed_time(&_vehicle_local_position.timestamp) < 1_s) {
+	launch_detection_status_s launch_detection_status{};
+	_launch_detection_status_sub.copy(&launch_detection_status);
+
+	// force the landed state to stay landed if we're currently in the catapult/hand-launch launch process. Detect that we are in this state
+	// by checking if the last publication of launch_detection_status is less than 0.5s old, and we're not yet in the flying state.
+	if (_landed_hysteresis.get_state() &&  hrt_elapsed_time(&launch_detection_status.timestamp) < 500_ms
```
[the remark this answers is not in the document]
Comment (tstastny) [reply src/modules/land_detector/FixedwingLandDetector.cpp:68]: nvm

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 4]
```diff
@@ -2259,6 +2259,8 @@ FixedwingPositionControl::Run()
 
 			if (_vehicle_land_detected_sub.update(&vehicle_land_detected)) {
 				_landed = vehicle_land_detected.landed;
+
+				if (_launch_detected) {_landed = false;}
 			}
 		}
```
[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:2265]: >can the outer if (_vehicle_land_detected_sub.updated()) be removed and just use the if (_vehicle_land_detected_sub.update(&vehicle_land_detected)) ?

I think you're totally right. I will propose that in a follow up.

[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:2265]: >Also - was this a case you saw were land detector had not yet negated the land state, but launch was already detected? maybe could add a short comment on the line here just to clarify the purpose.

That's a hack that I forgot to remove, now I handle it cleanly in the land detector itself (prevent to invalidate landed state if still in launch state machine).

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.hpp [hunk 5]
```diff
@@ -307,17 +309,15 @@ class FixedwingPositionControl final : public ModuleBase<FixedwingPositionContro
...
-
 	// true if a launch, specifically using the launch detector, has been detected
 	bool _launch_detected{false};
 
 	// [deg] global position of the vehicle at the time launch is detected (using launch detector)
 	Vector2d _launch_global_position{0, 0};
 
+	// [rad] current vehicle yaw at the time the launch is detected
+	float _launch_current_yaw{0.f};
+
```
[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.hpp:320]: _launch_current_yaw can actually should also be used for the runway takeoff. So not so sure if there is much (in terms of readability) to gain from structs.

File: src/modules/fw_pos_control_l1/fw_pos_control_l1_params.c [hunk 6]
```diff
@@ -1111,3 +1111,11 @@ PARAM_DEFINE_INT32(FW_LND_ABORT, 3);
...
  */
 PARAM_DEFINE_FLOAT(FW_WIND_ARSP_SC, 0.f);
+
+/**
+ * Launch detection
+ *
+ * @boolean
+ * @group FW Launch detection
+ */
+PARAM_DEFINE_INT32(LAUN_ALL_ON, 0);
```
[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/fw_pos_control_l1/fw_pos_control_l1_params.c:1121]: Done, now named LAUN_DECTN_ON

File: src/modules/fw_pos_control_l1/launchdetection/LaunchDetector.h [hunk 7]
```diff
@@ -33,51 +33,81 @@
...
-	LaunchDetectionResult getLaunchDetected();
-	bool launchDetectionEnabled() { return _param_laun_all_on.get(); }
+	/**
+	 * @brief Updates the state machine based on the current vehicle condition.
+	 *
+	 * @param dt Time step [us]
+	 * @param accel_x Measured acceleration in body x [m/s/s]
+	 * @param mavlink_log_pub
+	 */
+	void update(const float dt, float accel_x,  orb_advert_t *mavlink_log_pub);
```
[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/fw_pos_control_l1/launchdetection/LaunchDetector.h:71]: done.

File: src/modules/fw_pos_control_l1/launchdetection/LaunchDetector.h [hunk 8]
```diff
@@ -33,51 +33,81 @@
...
+	/**
+	 * Current state of the launch detection state machine [launch_detection_status_s::launch_detection_state]
+	 */
+	uint _state{launch_detection_status_s::STATE_WAITING_FOR_LAUNCH};
 
 	DEFINE_PARAMETERS(
-		(ParamBool<px4::params::LAUN_ALL_ON>) _param_laun_all_on
+		(ParamFloat<px4::params::LAUN_CAT_A>) _param_laun_cat_a,
+		(ParamFloat<px4::params::LAUN_CAT_T>) _param_laun_cat_t,
+		(ParamFloat<px4::params::LAUN_CAT_MDEL>) _param_laun_cat_mdel
```
[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/fw_pos_control_l1/launchdetection/LaunchDetector.h:110]: Right, I changed all class variables of the LaunchDetector to align with google style now.

File: src/modules/fw_pos_control_l1/launchdetection/LaunchDetector.cpp [hunk 9]
```diff
@@ -32,91 +32,95 @@
...
-	_activeLaunchDetectionMethodIndex = -1;
-}
-
-void LaunchDetector::update(const float dt, float accel_x)
-{
-	if (launchDetectionEnabled()) {
-		for (const auto launchMethod : _launchMethods) {
-			launchMethod->update(dt, accel_x);
+		/* Inform user that launchdetection is running every 4s */
+		if (_launchDetectionRunningInfoDelay >= 4.f) {
```
[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/fw_pos_control_l1/launchdetection/LaunchDetector.cpp:57]: good idea, done.

File: src/modules/fw_pos_control_l1/launchdetection/LaunchDetector.cpp [hunk 10]
```diff
@@ -32,91 +32,95 @@
...
-					_activeLaunchDetectionMethodIndex = i; // from now on only check this method
-					return _launchMethods[i]->getLaunchDetected();
+		/* Detect a acceleration that is longer and stronger as the minimum given by the params */
+		if (accel_x > _param_laun_cat_a.get()) {
+			_launchDetectionDelayCounter += dt;
+
+			if (_launchDetectionDelayCounter > _param_laun_cat_t.get()) {
+				if (_param_laun_cat_mdel.get() > 0.f) {
+					_state = launch_detection_status_s::STATE_LAUNCH_DETECTED_DISABLED_MOTOR;
+					mavlink_log_info(mavlink_log_pub, "Launch detected: enable control, waiting %8.1fs until full throttle\t",
```
[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/fw_pos_control_l1/launchdetection/LaunchDetector.cpp:70]: Named it " until throttling up" now (as the motors can already be enabled before, just at idle)

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (tstastny) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.hpp:320]: kk, sounds reasonable.

**Comment by sfuhrer** [comment]
rebased on base branch

**Review by tstastny — DISMISSED** [review]

**Comment by sfuhrer** [comment]
Did lots of hand-launch auto takeoffs today without issues. Only found the warning messages during the launch logic quite annoying, changed it in ac49bccd808f9679076e7f2057772b3cde1d6b78 to infos.

**Review by Jaeyoung-Lim — APPROVED** [review]
Couldn't find anything that I could find to nitpick :smile:

File: src/modules/fw_pos_control_l1/fw_pos_control_l1_params.c [hunk 11]
```diff
@@ -1093,3 +1110,15 @@ PARAM_DEFINE_INT32(FW_LND_ABORT, 3);
...
+ * FW Launch detection
+ *
+ * Enables automatic launch detection based on measured acceleration. Use for hand- or catapult-launched vehicles.
+ * Only available for fixed-wing vehicles.
+ * Not compatible with runway takeoff.
+ *
+ * @boolean
+ * @group FW Launch detection
+ */
+PARAM_DEFINE_INT32(LAUN_DETCN_ON, 0);
```
Comment (dagar) [inline src/modules/fw_pos_control_l1/fw_pos_control_l1_params.c:1124]: FW prefix?

**Comment by junwoo091400** [comment]
After another rebase onto `main`, the diff should be better 🙌, since https://github.com/PX4/PX4-Autopilot/pull/20537 got merged

File: src/modules/fw_pos_control_l1/fw_pos_control_l1_params.c (hunk 11 above)
Comment (sfuhrer) [reply src/modules/fw_pos_control_l1/fw_pos_control_l1_params.c]: Alright, renamed to FW_LAUN_DCTN_ON.
And to be consistent also changed the other 3 LAUN_* params. Incl. removing the CAT from the name, as it's a generic launch detector, not limited to work on a catapult.

**Comment by sfuhrer** [comment]
@dagar @tstastny As the CI was failing due to flash being over 100%, I've removed some outdated param translation and the legacy mavlink_log_pub from the FW position controller to safe some flash. The param translation trimming saves 0.05% flash (v5x), the mavlink_log_pub removing another 0.06%. 
Also rebased on main, and I think that alone would in this case already have been enough. Up to you if you want to add some extra buffer or if we should keep the param translation and mavlink_log_pub.

**Review by tstastny — APPROVED** [review]

File: src/modules/fw_pos_control_l1/launchdetection/LaunchDetector.h [hunk 12]
```diff
@@ -33,51 +33,83 @@
...
+	/**
+	 * Current state of the launch detection state machine [launch_detection_status_s::launch_detection_state]
+	 */
+	uint state_{launch_detection_status_s::STATE_WAITING_FOR_LAUNCH};
 
 	DEFINE_PARAMETERS(
-		(ParamBool<px4::params::LAUN_ALL_ON>) _param_laun_all_on
+		(ParamFloat<px4::params::FW_LAUN_AC_THLD>) _param_laun_cat_a,
+		(ParamFloat<px4::params::FW_LAUN_AC_T>) _param_laun_cat_t,
+		(ParamFloat<px4::params::FW_LAUN_MOT_DEL>) _param_laun_cat_mdel
```
[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/fw_pos_control_l1/launchdetection/LaunchDetector.h]: okay changed it and amended it to the last commit.

**Review by tstastny — APPROVED** [review]

**Review by tstastny — APPROVED** [review]

**Comment by Antiheavy** [comment]
related: https://github.com/PX4/PX4-Autopilot/pull/13797

**Pull request merged**